File: 재귀 & 깊이,넓이 우선탐색/미로탐색.cpp
#include <stdio.h>
using namespace std;

int map[8][8];
int dx[4] = {-1, 0, 1, 0};
int dy[4] = {0, 1, 0, -1};
int cnt;

void dfs(int a, int b){
	if(a == 7 && b == 7) cnt++;
	else{
		int x, y, i;
		for(i = 0; i < 4; i++){
			x = a + dx[i];
			y = b + dy[i];
			if(x >= 1 && x <= 7 && \
			   y >= 1 && y <= 7 && \
			   map[x][y] == 0){
				map[x][y] = 1;
				dfs(x, y);
				map[x][y] = 0;   	
			}
		}
	}
}

int main(void){
	// freopen("input.txt", "r", stdin);
	int i, j;
	
	for(i = 1; i <= 7; i++){
		for(j = 1; j <= 7; j++){
			scanf("%d", &map[i][j]);	
		}
	}
	
	map[1][1] = 1;
	dfs(1, 1);
	printf("%d", cnt);
	
	return 0;
}

File: DFS, BFS, 그래프/미로탐색.py
import sys
# sys.stdin = open("input.txt", 'r')

def dfs(i, j):
    global cnt
    if i == 6 and j == 6:
        cnt += 1
    else:
        for k in range(4):
            x = i + dx[k]
            y = j + dy[k]
            if 0 <= x <= 6 and \
               0 <= y <= 6 and \
               board[x][y] == 0:
                board[x][y] = 1
                dfs(x, y)
                board[x][y] = 0

if __name__ == "__main__":
    board = [list(map(int, input().split())) for _ in range(7)]
    dx = [-1, 0, 1, 0]
    dy = [0, 1, 0, -1]

    cnt = 0
    board[0][0] = 1
    dfs(0, 0)
    print(cnt)
